### DISCUSSION TITLE
Go API: config.Config: difference between AttachStd{in,out,err} and OpenStdin?

### DISCUSSION THREAD

**Discussion by thediveo** [moby/moby#47140] [Q&A]
Using the Docker Go client API, I want to create a new container and then attach to it in order to receive the container's (combined) output and optionally send it my input.

Reading https://docs.docker.com/engine/api/v1.43/#tag/Container/operation/ContainerCreate I'm still left in the dark:
1. what exactly does [`OpenStdin`](https://github.com/moby/moby/blob/c5187380d8b066bde9aff3d190ce621a67849f9b/api/types/container/config.go#L62) do? Which `stdin` is opened here, the one inside the container, or the one that the client process just uses?
2. even if I don't set `AttachStdin`, `AttachStdout` and `AttachStderr`, I still get input and output when I next do the `ContainerAttach` API call. So what are these fields actually doing when creating a container? And how do they relate to their similar named twins used in the `ContainerAttach` API call?

**Comment by thaJeztah** [comment]
Good question! These fields were added in the very beginning of Docker, and unfortunately many things from that time were poorly documented (the project was moving fast, and many things were still moving around).

I'd have to check the logic everywhere to exactly see how it's used. But started digging into Git history;

The `OpenStdIn` boolean was added before docker was open-sourced; there's no PR related to it, and was directly pushed to the repository in https://github.com/moby/moby/commit/7a50153c321ff96c16947b3bab9b8e385f2f2a5e;

> 'docker run -i' optionally opens stdin. 'docker attach' attaches to a running container (including stdin). 'docker run -t' allocates a tty (still buggy)

That commit already added some flags (`-i`, `-o`, `-e`) to attach STDIN, STDOUT and STDERR streams;

https://github.com/moby/moby/blob/7a50153c321ff96c16947b3bab9b8e385f2f2a5e/dockerd/dockerd.go#L526-L530

https://github.com/moby/moby/blob/7a50153c321ff96c16947b3bab9b8e385f2f2a5e/dockerd/dockerd.go#L572-L576


The `AttachStdin`, `AttachStdout`, and `AttachStderr` booleans were added in https://github.com/moby/moby/commit/c04af2a330991fcd52bcce213bcb863cff95d378, which was part of;

- https://github.com/moby/moby/pull/315
- to address: https://github.com/moby/moby/issues/234

At a quick glance, I wonder if those fields were meant to _replace_ the `OpenStdIn` field (but providing more granular options), but I need to have a more in-depth look (perhaps one of the other maintainers knows by head).

We should definitely improve the docs for those fields (or perhaps one of them is redundant, but never removed :joy:)

**Reply by thediveo** [reply]
Thank you very much for this first insight! Larger projects tend to have their history anyway, so this somehow confirms my gut feeling that I'm dealing with things that grew over time and could not be easily refactored as there are already enough things downstream still depending on them.

Researching more, I came across this Docker community forums post: [Understanding docker run --attach option](https://forums.docker.com/t/understanding-docker-run-attach-option/134337), and [rimelek's](https://forums.docker.com/u/rimelek) super helpful `docker run ...`/`docker inspect ...` experiments. Based on his findings I drew up the following condensed table: 

| Options | `AttachStdin` | `AttachStdout`/ `AttachStderr` | `OpenStdin` | `StdinOnce` | `Tty` |
| ---- | ---- | ---- | ---- | ---- | ---- |
|  | `false` | `true`/`true` | `false` | `false` | `false` |
| `--attach stdin` | `true` | `false`/`false` | `false` | `false` | `false` |
| `--attach stderr` | `false` | `false`/`true` | `false` | `false` | `false` |
| `-i` | **`true`** | `true`/`true` | **`true`** | **`true`** | `false` |
| `-it` | **`true`** | `true`/`true` | **`true`** | **`true`** | **`true`** |
| `-i --attach stderr` | **`true`** | `false`/**`true`** | **`true`** | **`true`** | `false` |
| `-d` | `false` | `false`/`false` | **`true`** | `false` | `false` |

From what I could glance already from the `docker` CLI implementation for [`runRun`](https://github.com/docker/cli/blob/f18a476b6d240bafbaefb65e51f837eab9e02b94/cli/command/container/run.go#L92) (the `docker run` command implementation):
- the above settings are passed through the container config into the `ContainerCreate` API client call.
- if not running detached,  `AttachStdin`, `AttachStdout`, and `AttachStderr` are then additionally passed for attaching, if at least one of them has been set to `true`.

Now while the purpose of `AttachXXX` is clear to me for the attach API call, I'm struggling to understand their meaning for _creating_ the container in the first place. Might they be used in container creation to allocate pipes/ttys for the fds 0, 1, 2?

That is, does "attach" for container creation mean something different from what it means for the attach operation?
- "attach" when creating container: allocate pipes/tty on the daemon host and _attach_ fds 0, 1, 2 inside the container to them.
- "attach" then attaching: stream the specified pipes/tty fds 0, 1, 2 (any combination of them as requested) from the daemon's side to the client side.

What probably is confusing me is that I can perfectly programmatically create the container with no attachment flags set, and then attach with streaming all three stdin, stdout, stderr correctly (setting `OpenStdin` and `StdinOnce` beforehand when creating the container though).

Comments, feedback, please?
